### PR TITLE
speed up redis listall interface

### PR DESCRIPTION
**Pull request by zhijian-pro**

### CODE REVIEW

File: pkg/object/filesystem_test.go [hunk 1]
```diff
@@ -64,6 +64,11 @@ func TestHDFS2(t *testing.T) {
 	testFileSystem(t, dfs)
 }
 
+func TestRedis2(t *testing.T) {
```
Comment (davies) [inline pkg/object/filesystem_test.go:67]: Redis don't have file system sematics, don't need to test it.

File: pkg/object/redis.go [hunk 2]
```diff
@@ -79,13 +79,13 @@ func (t *redisStore) ListAll(prefix, marker string) (<-chan Object, error) {
...
 	for {
 		// FIXME: this will be really slow for many objects
-		keys, c, err := t.rdb.Scan(context.TODO(), cursor, "*", 1000).Result()
+		keys, c, err := t.rdb.Scan(context.TODO(), cursor, prefix+"*", 1000).Result()
 		if err != nil {
 			logger.Warnf("redis scan error, coursor %d: %s", cursor, err)
 			return nil, err
 		}
 		for _, key := range keys {
-			if key > marker && strings.HasPrefix(key, prefix) {
```
Comment (davies) [inline pkg/object/redis.go:88]: We can use Len and Pipeline to speed up it

File: pkg/object/redis.go [hunk 3]
```diff
@@ -94,19 +94,30 @@ func (t *redisStore) ListAll(prefix, marker string) (<-chan Object, error) {
 		}
 		cursor = c
 	}
+
 	sort.Strings(keyList)
-	now := time.Now()
+	p := t.rdb.Pipeline()
```
Comment (davies) [inline pkg/object/redis.go:99]: 一个 batch 的 数量应该有个限制

File: pkg/object/redis.go [hunk 4]
```diff
@@ -94,19 +94,30 @@ func (t *redisStore) ListAll(prefix, marker string) (<-chan Object, error) {
...
 	for _, key := range keyList {
-		data, err := t.rdb.Get(c, key).Bytes()
-		if err != nil {
-			if err == redis.Nil {
-				continue
+		p.Get(c, key)
```
Comment (davies) [inline pkg/object/redis.go:101]: 用 Len

File: pkg/utils/utils.go [hunk 5]
```diff
@@ -108,3 +108,21 @@ func StringContains(s []string, e string) bool {
 	}
 	return false
 }
+
+func SliceSplitWithMapFunc(srcSlice []string, batch int, f func(subSlice []string) error) error {
```
Comment (davies) [inline pkg/utils/utils.go:112]: this one is not needed

File: pkg/object/redis.go [hunk 6]
```diff
@@ -94,19 +95,45 @@ func (t *redisStore) ListAll(prefix, marker string) (<-chan Object, error) {
...
 			return nil, err
 		}
 
-		// FIXME: mtime
-		objs <- &obj{key, int64(len(data)), now, strings.HasSuffix(key, "/")}
+		now := time.Now()
+		for idx, cmd := range cmds {
+			if intCmd, ok := cmd.(*redis.IntCmd); ok {
+				size, err := intCmd.Result()
+				if err != nil {
```
Comment (davies) [inline pkg/object/redis.go:121]: Will err be Nil?

Comment (zhijian-pro) [reply pkg/object/redis.go:121]: err will not be nil but size will be 0 when key does not exist

**Pull request merged**